### ISSUE TITLE
Cast values to the proper type and use `Ash.Type.equal?`

### ISSUE THREAD

**Issue by moxley** [ash-project/ash#665]
**Describe the bug**

Inconsistent behavior in `Ash.Changeset.manage_relationship()` with respect to the data types in its 3rd argument (attrs).

This issue was described in the Discord Support channel, under the title "Using manage_relationship to delete a related record".

**To Reproduce**

Below is a test that demonstrates updating and deleting a Note from a Member. In this particular application, a `Member` has_many `Note`s. The underlying `notes` table uses an integer primary key. When defining actions for `Member` that manage notes for the `Member`, I want to define a string `argument` for the Note ID, so that when the argument is exposed in a web API as a string instead of an integer. This follows good web API design, which is to prefer string IDs over integer ones, even if the underlying data type of the ID is an integer.

This test shows that a string Note ID works for the "update note" case, but not for the "delete note" case:

```elixir
test "delete note", ctx do
  group = ctx.group

  actor = insert_member(group_id: group.id, name: "Actor", active: true, roles: ["members"])
  member = insert_member(group_id: group.id, name: "Member", active: true)
  note = insert_note(group_id: group.id, actor: actor, body: "new note", type: "member-internal", member_id: member.id)

  # String note.id works for `manage_relationship(:notes, [note_attrs], type: :append)`, below
  note_attrs = %{id: to_string(note.id), body: "updated note"}

  member =
    member
    |> Ash.Changeset.for_update(:update, %{}, actor: actor)
    |> Ash.Changeset.manage_relationship(:notes, [note_attrs], type: :append)
    |> MyApp.AshApi.update!()
    |> MyApp.AshApi.load!(:notes, authorize?: false)

  assert [updated_note] = member.notes
  assert updated_note.body == "updated note"

  # String note.id fails for `manage_relationship(:notes, [note_attrs], type: {:destroy, _})`, below
  note_attrs = %{id: to_string(note.id)}

  member =
    member
    |> Ash.Changeset.for_update(:update, %{}, actor: actor)
    |> Ash.Changeset.manage_relationship(:notes, [note_attrs], on_match: {:destroy, :destroy})
    |> MyApp.AshApi.update!()
    |> MyApp.AshApi.load!(:notes, authorize?: false)

  # Fails. Note is not deleted
  assert member.notes == []
end
```

**Expected behavior**

When I define a string `argument` that represents a record ID whose underlying type is an integer, I expect `Ash.Changeset.manage_relationship(_, _, _, on_match: {:destroy, _})` to work with that argument value just as well as `Ash.Changeset.manage_relationship(_, _, _, type: :append)` does.

**Runtime**
 - Elixir version 1.14
 - Erlang version 25.2
 - OS MacOS Silicon
 - Ash version 2.9.29
 - any related extension versions

**Additional context**

This issue was described in the Discord Support channel, under the title "Using manage_relationship to delete a related record".

**Comment by zachdaniel** [comment]
Thanks for the detailed report. This one is a bit more involved for the fix, so I'll need to take a stab at it this weekend or Monday.

**Comment by moxley** [comment]
> Thanks for the detailed report. This one is a bit more involved for the fix, so I'll need to take a stab at it this weekend or Monday.

No rush @zachdaniel. I have a good workaround in place.

**Comment by moxley** [comment]
Are you the same person who texted me? The person who texted me is coming
to get it.

**Comment by ghalestrilo** [comment]
@zachdaniel with the correct pointers I may be able to help here.

**Comment by zachdaniel** [comment]
👋 So what needs to happen is that we need to change this behavior:

https://github.com/ash-project/ash/blob/main/lib/ash/actions/managed_relationships.ex#L1440

Currently it's using `Comp.equal?`. but what we need to do is cast each input using `Ash.Type.cast_input`, and to do that we need to pass down the type along with the key we are checking. Not the most in-depth explanation. LMK if that is enough to go on :)

**Comment by ghalestrilo** [comment]
Enough to give it a go :) Thanks

**Comment by ghalestrilo** [comment]
@moxley I tried replicating your test , but it seems to pass 🤔 
Does thi code do anything different than your example?

```elixir
test "uses deep equality to check for deletion", ctx do
  actor =
    Member
    |> Changeset.new(%{name: "Actor", active: true, roles: ["members"]})
    |> Api.create!()

  member =
    Member
    |> Changeset.new(%{name: "Member", active: true})
    |> Api.create!()

  note =
    Note
    |> Changeset.new(%{body: "new note", type: "member-internal"})
    |> Changeset.manage_relationship(:member, member)
    |> Api.create!()

  # This triggers an error in `assert updated_note.body == "updated note"`
  # note =
  #   Note
  #   |> Changeset.new(%{body: "new note", type: "member-internal", member_id: member.id})
  #   |> Api.create!()

  note_attrs = %{id: to_string(note.id), body: "updated note"}

  member =
    member
    |> Changeset.for_update(:update, %{}, actor: actor)
    |> Changeset.manage_relationship(:notes, [note_attrs], type: :append)
    |> Api.update!()
    |> Api.load!(:notes, authorize?: false)

  assert [updated_note] = member.notes
  assert length(member.notes) == 1
  assert updated_note.body == "updated note"

  note_attrs = %{id: to_string(note.id)}

  member =
    member
    |> Changeset.for_update(:update, %{}, actor: actor)
    |> Changeset.manage_relationship(:notes, [note_attrs], on_match: {:destroy, :destroy})
    |> Api.update!()
    |> Api.load!(:notes, authorize?: false)

  assert member.notes == []
end
```

Using the following implementation for `Note` and `Member`

```elixir
defmodule Note do
  @moduledoc false
  use Ash.Resource, data_layer: Ash.DataLayer.Ets

  ets do
    private?(false)
  end

  actions do
    defaults [:read, :create, :update, :destroy]
  end

  attributes do
    uuid_primary_key :id
    attribute :body, :string
    attribute :type, :string
    attribute :member_id, :string
  end

  relationships do
    belongs_to :member, Member, source_attribute: :member_id
  end
end

defmodule Member do
  @moduledoc false
  use Ash.Resource, data_layer: Ash.DataLayer.Ets

  ets do
    private?(false)
  end

  actions do
    defaults [:read, :create, :update, :destroy]
  end

  attributes do
    uuid_primary_key :id
    attribute :name, :string
    attribute :roles, {:array, :string}
    attribute :active, :boolean
  end

  relationships do
    has_many :notes, Note, destination_attribute: :member_id
  end
end
```

**Comment by moxley** [comment]
Hi @ghalestrilo It's probably not worth trying to solve this. I'm closing this issue.

**Comment by vonagam** [comment]
Oh, then mine issue #877 is a duplicate one. It's worth fixing, it's just not urgent, I guess.

**Comment by StephanH90** [comment]
This has been fixed as of 2.21.7 see [00eba86](https://github.com/ash-project/ash/commit/00eba8672c35015b0a529cc2a389cdb221b65c53)
